worker/firewaller/firewaller.go

```diff
@@ -1296,17 +1314,59 @@ func (rd *remoteRelationData) watchLoop() error {
 		}
 	}
 
+	if rd.isOffer {
+		// On the offering side we watch the local model for ingress changes
+		// which will have been published from the consuming model.
+		rd.ingressAddressWatcherFunc = func() (watcher.StringsWatcher, error) {
```

Comment:
I would find this method a lot easier to follow if `remoteRelationData` had two methods `watchIngressAddresses` and `watchEgressAddresses`, rather than the ingress/egress funcs. As far as I can see the closures here only use fields of rd and they don't get called so often that it's a problem to make the checks for isOffer and provider/requirer each call.